### PR TITLE
Bug report: Incorrect ranges in key condition

### LINKED ISSUE #13248: Assertion "Cannot read out of marks range"
See https://clickhouse-test-reports.s3.yandex.net/13238/6f690b7c0da14335dc28ae29adfc0b40038f6fe6/fuzzer/fuzzer.log#fail1

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
To be added Fixes: #13248

### CODE REVIEW

**Comment by alesapin** [comment]
Actually, we get correct key condition `Key condition: (column 0 in [4294967280, +inf)), (column 0 in (-inf, 0]), and` And it's always false for all ranges, but for some reason `mayBeTrueAfter` https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/KeyCondition.cpp#L1372-L1378 return true.

**Comment by alesapin** [comment]
Actually our code works badly with impossible KeyConditions. In this case, we have condition:
```
(column 0 in [4294967280, +inf)), (column 0 in (-inf, 0]), and
            >>>>>first<<<<<                  >>>>>second<<<<<
```
Which is obviously always false for any interval. But for example, for interval `[0, +inf)` our code will consider that it intersects first interval (on ` [4294967280, +inf)`) and second interval in point `0`. After that, we will evaluate `AND` (https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/MergeTree/KeyCondition.cpp#L1304-L1311)[link] and yes, in both ranges our key condition "can be true" and we will try to read something from disk (actually full column). 

I think we can simplify KeyCondition's RPN after it was parsed from `WHERE` expression, and get rid of similar bugs. For example, if we have `OR` expression and left interval contains the right interval than we can replace this element with the left interval. Or if a left interval of `AND` expression doesn't intersect the right interval than we can replace `AND` with `ALWAYS_FALSE` (this case). But it can be quite complicated for a multidimensional KeyConditions.

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -285,6 +285,53 @@ static const std::map<std::string, std::string> inverse_relations = {
...
+            {
+                if (arg1.function == RPNElement::ALWAYS_TRUE && arg2.function == RPNElement::ALWAYS_TRUE)
+                    rpn_stack.back() = RPNElement(RPNElement::ALWAYS_TRUE);
+                else if (arg1.function == RPNElement::ALWAYS_FALSE || arg2.function == RPNElement::ALWAYS_FALSE)
+                    rpn_stack.back() = RPNElement(RPNElement::ALWAYS_FALSE);
+                else if (arg1.function == RPNElement::ALWAYS_TRUE)
+                    rpn_stack.back() = arg2;
+                else if (arg2.function == RPNElement::ALWAYS_TRUE)
+                    rpn_stack.back() = arg1;
+                else if (arg1.function == RPNElement::FUNCTION_NOT_IN_RANGE && arg2.function == RPNElement::FUNCTION_NOT_IN_RANGE && arg1.range.intersectsRange(arg2.range))
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/KeyCondition.cpp:311]: This looks incorrect.
```
range1:          *****
range2:             *****
not range1: *****     *****
not range2: ********     **
not both:   *****        **
```

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 2]
```diff
@@ -285,6 +285,53 @@ static const std::map<std::string, std::string> inverse_relations = {
 };
 
 
+void KeyCondition::optimizeRPN()
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/KeyCondition.cpp:288]: The idea of RPN optimization is :+1:

**Comment by alesapin** [comment]
Closed in favor to #14225.

**Pull request closed without merging**